In [ ]:
!pip install neuralforecast

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.6/170.6 kB 2.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 776.9/776.9 kB 16.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.6/62.6 MB 12.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 413.4/413.4 kB 25.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.1/806.1 kB 38.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.7/101.7 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.6/230.6 kB 7.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 4.1 MB/s eta 0:00:00


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math
import numpy as np
import torch
from typing import Optional, Union, Tuple


import torch.nn.functional as F
from torch.distributions import Distribution
from torch.distributions import Bernoulli, Normal, StudentT, Poisson, NegativeBinomial

from torch.distributions import constraints


from neuralforecast import NeuralForecast
from neuralforecast.models import PatchTST
from neuralforecast.tsdataset import TimeSeriesDataset
from neuralforecast.utils import AirPassengers, AirPassengersPanel, AirPassengersStatic
from neuralforecast.losses.numpy import rmse, mape
from pandas.tseries.holiday import USFederalHolidayCalendar
from pandas.tseries.offsets import CustomBusinessDay
from neuralforecast.losses.pytorch import MQLoss, DistributionLoss, MSE, MAE

In [ ]:
from numpy.random import seed
from random import randrange

In [ ]:
df = pd.read_excel("SP100_vol (clean).xlsx", index_col="Date",parse_dates=True)
columns=df.columns
df_static=pd.read_excel("Static_df.xlsx")
n_series=len(df.columns)
train = df.iloc[:int(len(df['NKE'])*0.7)]
validation_length=int(len(df['NKE'])*0.7)-int(len(df['NKE'])*0.5)
train=train.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

In [ ]:
n_inputs = [3,5,10,15,21,42,84]
hidden_size = [60,90,120,150,180]
epochs=[50,100,150,250,350,450,550,650,750]
learning_rate=[0.0005,0.0001,0.00005,0.00001]
num_lr_decays=[5,3,2,1]
scaler_type=["robust","standard",'minmax']
losses=[MSE(),MAE(),MQLoss(level=[80, 90]),DistributionLoss(distribution='StudentT', level=[80, 90])]
n_head=[6,9,12,15,18]
decoder_input_size_multiplier=[0.3,0.4,0.5,0.6,0.7]
dropout=[0,0.2,0.3,0.4]
patch_len=[4,6,8,10,12]
stride=[6,7,8,9,10]
encoder_layers = [1,2,3,4,5]

# Hyperparameter Search

In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 63510272


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 78613239


Epoch: 650 | Inputs: 21 | hidden_size: 0 | Loss: 0 | Scaler Type: minmax | attn_dropout: 0.4 | encoder_layers: 5
Learning Rate: 1 | head_dropout: 0.4 | Dropout: 0.3 | n_head: 6 | patch_len: 0
RMSE: 0.0025523732638995118 | QLIKE: 0.022927532831586217 | MAE: 0.0014952542393252284 | MAPE: 0.15102629083124328


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 93277816


Epoch: 750 | Inputs: 3 | hidden_size: 1 | Loss: 1 | Scaler Type: standard | attn_dropout: 0.4 | encoder_layers: 2
Learning Rate: 2 | head_dropout: 0.4 | Dropout: 0.4 | n_head: 9 | patch_len: 1
RMSE: 0.0025883291882798642 | QLIKE: 0.02371765674842075 | MAE: 0.001458495708795374 | MAPE: 0.14389524625292424


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 30374290


Epoch: 100 | Inputs: 5 | hidden_size: 0 | Loss: 3 | Scaler Type: standard | attn_dropout: 0.2 | encoder_layers: 5
Learning Rate: 2 | head_dropout: 0.2 | Dropout: 0.4 | n_head: 6 | patch_len: 0
RMSE: 0.0025627059251489685 | QLIKE: 0.023509796011454764 | MAE: 0.0014791490926394694 | MAPE: 0.145301477228466


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 55127851


Epoch: 100 | Inputs: 15 | hidden_size: 0 | Loss: 2 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 3
Learning Rate: 0 | head_dropout: 0.3 | Dropout: 0.3 | n_head: 6 | patch_len: 0
RMSE: 0.0025603030128472483 | QLIKE: 0.023556079186776446 | MAE: 0.0015041798288661515 | MAPE: 0.14807221479304203


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 450 | Inputs: 3 | hidden_size: 0 | Loss: 1 | Scaler Type: minmax | attn_dropout: 0.2 | encoder_layers: 1
Learning Rate: 3 | head_dropout: 0.2 | Dropout: 0.3 | n_head: 6 | patch_len: 0
RMSE: 0.0025997283822753498 | QLIKE: 0.023873242116607483 | MAE: 0.0014405921637271606 | MAPE: 0.1425073072475016


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 45278987


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 97204381


Epoch: 100 | Inputs: 21 | hidden_size: 0 | Loss: 0 | Scaler Type: robust | attn_dropout: 0.3 | encoder_layers: 5
Learning Rate: 0 | head_dropout: 0.3 | Dropout: 0.2 | n_head: 6 | patch_len: 0
RMSE: 0.0025361812774718547 | QLIKE: 0.02287165542426189 | MAE: 0.0015256823906616963 | MAPE: 0.15325221201738753


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 14253827


Epoch: 350 | Inputs: 84 | hidden_size: 1 | Loss: 2 | Scaler Type: standard | attn_dropout: 0.4 | encoder_layers: 4
Learning Rate: 1 | head_dropout: 0.4 | Dropout: 0.2 | n_head: 9 | patch_len: 1
RMSE: 0.0025717717140244147 | QLIKE: 0.023840557295272118 | MAE: 0.0014784508144809322 | MAPE: 0.14347199273562514


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 42751648


Epoch: 150 | Inputs: 15 | hidden_size: 0 | Loss: 3 | Scaler Type: minmax | attn_dropout: 0.4 | encoder_layers: 2
Learning Rate: 3 | head_dropout: 0.4 | Dropout: 0.4 | n_head: 6 | patch_len: 2
RMSE: 0.002668495080158738 | QLIKE: 0.026105536624289348 | MAE: 0.0015633427082462528 | MAPE: 0.15449466316813595


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 19030040


Epoch: 250 | Inputs: 5 | hidden_size: 2 | Loss: 3 | Scaler Type: robust | attn_dropout: 0 | encoder_layers: 4
Learning Rate: 2 | head_dropout: 0 | Dropout: 0.2 | n_head: 12 | patch_len: 0
RMSE: 0.0025986867561474825 | QLIKE: 0.023822853098915853 | MAE: 0.0015588771984963758 | MAPE: 0.15159175346234297


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 350 | Inputs: 15 | hidden_size: 4 | Loss: 2 | Scaler Type: minmax | attn_dropout: 0.4 | encoder_layers: 2
Learning Rate: 0 | head_dropout: 0.4 | Dropout: 0.3 | n_head: 18 | patch_len: 2
RMSE: 0.00254056277794981 | QLIKE: 0.023174892497986682 | MAE: 0.0014739689872967832 | MAPE: 0.14376088593078926


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 126692174


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 42620035


Epoch: 350 | Inputs: 15 | hidden_size: 4 | Loss: 0 | Scaler Type: robust | attn_dropout: 0.3 | encoder_layers: 5
Learning Rate: 0 | head_dropout: 0.3 | Dropout: 0 | n_head: 18 | patch_len: 4
RMSE: 0.002538096380694767 | QLIKE: 0.02283947763996506 | MAE: 0.0014942533802285523 | MAPE: 0.14998241150142574


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 115357812


Epoch: 450 | Inputs: 5 | hidden_size: 0 | Loss: 3 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 5
Learning Rate: 3 | head_dropout: 0.3 | Dropout: 0 | n_head: 6 | patch_len: 3
RMSE: 0.0025618454983810358 | QLIKE: 0.023597728507002484 | MAE: 0.0014869615879240059 | MAPE: 0.14472919650057967


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 39700876


Epoch: 350 | Inputs: 3 | hidden_size: 0 | Loss: 2 | Scaler Type: standard | attn_dropout: 0 | encoder_layers: 5
Learning Rate: 1 | head_dropout: 0 | Dropout: 0.3 | n_head: 6 | patch_len: 2
RMSE: 0.0026235683488713196 | QLIKE: 0.023668397225805337 | MAE: 0.001565068983961383 | MAPE: 0.15543866839977438


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 49961512


Epoch: 350 | Inputs: 3 | hidden_size: 2 | Loss: 2 | Scaler Type: standard | attn_dropout: 0.4 | encoder_layers: 3
Learning Rate: 0 | head_dropout: 0.4 | Dropout: 0 | n_head: 12 | patch_len: 1
RMSE: 0.002571113069296258 | QLIKE: 0.02367193763942682 | MAE: 0.0014554166188171834 | MAPE: 0.14259962865608947


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 150 | Inputs: 3 | hidden_size: 4 | Loss: 1 | Scaler Type: standard | attn_dropout: 0 | encoder_layers: 3
Learning Rate: 2 | head_dropout: 0 | Dropout: 0.4 | n_head: 18 | patch_len: 1
RMSE: 0.0025914263778733463 | QLIKE: 0.023709140242357457 | MAE: 0.0015306022505306826 | MAPE: 0.14912474468885392


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 101682327


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 108666147


Epoch: 750 | Inputs: 84 | hidden_size: 3 | Loss: 3 | Scaler Type: robust | attn_dropout: 0.2 | encoder_layers: 4
Learning Rate: 1 | head_dropout: 0.2 | Dropout: 0 | n_head: 15 | patch_len: 0
RMSE: 0.002595681712005272 | QLIKE: 0.02424522256438349 | MAE: 0.0015270632211663118 | MAPE: 0.14874899020798205


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 24258898


Epoch: 350 | Inputs: 10 | hidden_size: 3 | Loss: 1 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 1
Learning Rate: 3 | head_dropout: 0.3 | Dropout: 0.4 | n_head: 15 | patch_len: 4
RMSE: 0.0025568691278811304 | QLIKE: 0.02333312547506533 | MAE: 0.0015220044330604562 | MAPE: 0.1506627014648355


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 49487314


Epoch: 350 | Inputs: 3 | hidden_size: 4 | Loss: 2 | Scaler Type: minmax | attn_dropout: 0.4 | encoder_layers: 3
Learning Rate: 1 | head_dropout: 0.4 | Dropout: 0.3 | n_head: 18 | patch_len: 1
RMSE: 0.002604127967465722 | QLIKE: 0.02361034673642883 | MAE: 0.0014720678037239618 | MAPE: 0.14641067725325022


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 119575181


Epoch: 650 | Inputs: 5 | hidden_size: 0 | Loss: 3 | Scaler Type: standard | attn_dropout: 0.2 | encoder_layers: 2
Learning Rate: 0 | head_dropout: 0.2 | Dropout: 0.3 | n_head: 6 | patch_len: 0
RMSE: 0.0025606978236830517 | QLIKE: 0.023416093816611878 | MAE: 0.0015004705030538692 | MAPE: 0.14641106413095833


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 650 | Inputs: 21 | hidden_size: 0 | Loss: 1 | Scaler Type: minmax | attn_dropout: 0.2 | encoder_layers: 2
Learning Rate: 3 | head_dropout: 0.2 | Dropout: 0.3 | n_head: 6 | patch_len: 1
RMSE: 0.0025576795524376995 | QLIKE: 0.023333768841070433 | MAE: 0.0015269686490010472 | MAPE: 0.1488013006216861


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 75887829


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 119932973


Epoch: 750 | Inputs: 42 | hidden_size: 1 | Loss: 2 | Scaler Type: minmax | attn_dropout: 0.2 | encoder_layers: 1
Learning Rate: 1 | head_dropout: 0.2 | Dropout: 0 | n_head: 9 | patch_len: 4
RMSE: 0.002551727680160319 | QLIKE: 0.02325903545104811 | MAE: 0.0015096502599094477 | MAPE: 0.1453899688496068


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 88782229


Epoch: 50 | Inputs: 5 | hidden_size: 3 | Loss: 1 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 4
Learning Rate: 2 | head_dropout: 0.3 | Dropout: 0 | n_head: 15 | patch_len: 2
RMSE: 0.002601930668581518 | QLIKE: 0.02351160360803599 | MAE: 0.0015361530417454814 | MAPE: 0.1540577932291325


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 57579602


Epoch: 750 | Inputs: 10 | hidden_size: 4 | Loss: 0 | Scaler Type: minmax | attn_dropout: 0.4 | encoder_layers: 3
Learning Rate: 3 | head_dropout: 0.4 | Dropout: 0.4 | n_head: 18 | patch_len: 1
RMSE: 0.002560845599699202 | QLIKE: 0.023091813059192865 | MAE: 0.00150760690745415 | MAPE: 0.1516763984517001


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 46481869


Epoch: 450 | Inputs: 21 | hidden_size: 2 | Loss: 2 | Scaler Type: minmax | attn_dropout: 0.4 | encoder_layers: 3
Learning Rate: 0 | head_dropout: 0.4 | Dropout: 0.3 | n_head: 12 | patch_len: 2
RMSE: 0.0025459066655683536 | QLIKE: 0.023319122082528152 | MAE: 0.0014602266171275724 | MAPE: 0.14221977775802397


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 750 | Inputs: 15 | hidden_size: 3 | Loss: 0 | Scaler Type: robust | attn_dropout: 0.2 | encoder_layers: 2
Learning Rate: 2 | head_dropout: 0.2 | Dropout: 0.2 | n_head: 15 | patch_len: 1
RMSE: 0.0025892102202141162 | QLIKE: 0.023965892596191404 | MAE: 0.0015813346821842412 | MAPE: 0.15858708841783992


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 124151285


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 62312283


Epoch: 50 | Inputs: 21 | hidden_size: 0 | Loss: 1 | Scaler Type: minmax | attn_dropout: 0.2 | encoder_layers: 4
Learning Rate: 0 | head_dropout: 0.2 | Dropout: 0 | n_head: 6 | patch_len: 2
RMSE: 0.0025591032182508446 | QLIKE: 0.023264914187726603 | MAE: 0.0015330110418004196 | MAPE: 0.14927636450165296


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 52090623


Epoch: 550 | Inputs: 84 | hidden_size: 1 | Loss: 2 | Scaler Type: standard | attn_dropout: 0.4 | encoder_layers: 1
Learning Rate: 0 | head_dropout: 0.4 | Dropout: 0.4 | n_head: 9 | patch_len: 3
RMSE: 0.002554459608860689 | QLIKE: 0.02338671380300938 | MAE: 0.0014751350150293103 | MAPE: 0.14337950062198737


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 25319499


Epoch: 150 | Inputs: 84 | hidden_size: 3 | Loss: 1 | Scaler Type: minmax | attn_dropout: 0.2 | encoder_layers: 2
Learning Rate: 3 | head_dropout: 0.2 | Dropout: 0.3 | n_head: 15 | patch_len: 3
RMSE: 0.002684951343219253 | QLIKE: 0.02625130663470146 | MAE: 0.0016504385932828863 | MAPE: 0.16339566492414637


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 25084942


Epoch: 350 | Inputs: 15 | hidden_size: 0 | Loss: 2 | Scaler Type: robust | attn_dropout: 0.2 | encoder_layers: 5
Learning Rate: 0 | head_dropout: 0.2 | Dropout: 0.4 | n_head: 6 | patch_len: 3
RMSE: 0.002568442072447495 | QLIKE: 0.023678670898367983 | MAE: 0.0015287143876436522 | MAPE: 0.1483228692488742


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 350 | Inputs: 15 | hidden_size: 3 | Loss: 2 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 3
Learning Rate: 0 | head_dropout: 0.3 | Dropout: 0 | n_head: 15 | patch_len: 4
RMSE: 0.002529506699827837 | QLIKE: 0.022942362295855584 | MAE: 0.0014807732587621629 | MAPE: 0.144198631314966


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 116946082


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 89009207


Epoch: 150 | Inputs: 84 | hidden_size: 4 | Loss: 3 | Scaler Type: robust | attn_dropout: 0.2 | encoder_layers: 5
Learning Rate: 2 | head_dropout: 0.2 | Dropout: 0 | n_head: 18 | patch_len: 3
RMSE: 0.0025587320894958237 | QLIKE: 0.023444046591734234 | MAE: 0.0015165514768282144 | MAPE: 0.14785765658271804


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 84237992


Epoch: 250 | Inputs: 84 | hidden_size: 1 | Loss: 2 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 2
Learning Rate: 2 | head_dropout: 0.3 | Dropout: 0.4 | n_head: 9 | patch_len: 3
RMSE: 0.0025742289887255325 | QLIKE: 0.02366583994126155 | MAE: 0.001525782790392922 | MAPE: 0.15181103524020234


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 57313573


Epoch: 250 | Inputs: 10 | hidden_size: 2 | Loss: 1 | Scaler Type: minmax | attn_dropout: 0.2 | encoder_layers: 3
Learning Rate: 3 | head_dropout: 0.2 | Dropout: 0.2 | n_head: 12 | patch_len: 4
RMSE: 0.0025576645207486762 | QLIKE: 0.02321889154865371 | MAE: 0.0015382553002006175 | MAPE: 0.15216792109457883


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 86940972


Epoch: 650 | Inputs: 84 | hidden_size: 0 | Loss: 0 | Scaler Type: minmax | attn_dropout: 0 | encoder_layers: 1
Learning Rate: 1 | head_dropout: 0 | Dropout: 0 | n_head: 6 | patch_len: 1
RMSE: 0.002581981524901075 | QLIKE: 0.02363002272611727 | MAE: 0.0016024680766991012 | MAPE: 0.1611725529154388


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 650 | Inputs: 10 | hidden_size: 3 | Loss: 3 | Scaler Type: minmax | attn_dropout: 0 | encoder_layers: 3
Learning Rate: 1 | head_dropout: 0 | Dropout: 0.3 | n_head: 15 | patch_len: 2
RMSE: 0.0026049252896022607 | QLIKE: 0.02391052561719365 | MAE: 0.001569140956236811 | MAPE: 0.15143743483217884


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  m1=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(5)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(5)
  a6=randrange(4)
  model = PatchTST(h=1, input_size=n_inputs[i],
             hidden_size=hidden_size[h], linear_hidden_size =int(hidden_size[h]*2),
           dropout=dropout[m],  fc_dropout=dropout[m], head_dropout=dropout[a6],
                   patch_len=patch_len[a2],stride=stride[a2],
           attn_dropout=dropout[a6], n_heads=n_head[h],
             encoder_layers=encoder_layers[a5],loss=losses[m1],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {h} | Loss: {m1} | Scaler Type: {scaler_type[o]} | attn_dropout: {dropout[a6]} | encoder_layers: {encoder_layers[a5]}")
  print(f"Learning Rate: {n} | head_dropout: {dropout[a6]} | Dropout: {dropout[m]} | n_head: {n_head[h]} | patch_len: {a2}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 75931618


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 31220686


Epoch: 550 | Inputs: 84 | hidden_size: 0 | Loss: 0 | Scaler Type: robust | attn_dropout: 0.4 | encoder_layers: 1
Learning Rate: 0 | head_dropout: 0.4 | Dropout: 0.4 | n_head: 6 | patch_len: 4
RMSE: 0.002566489174346347 | QLIKE: 0.023108994625726045 | MAE: 0.0014829551768005202 | MAPE: 0.14832645240230316


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 42914685


Epoch: 450 | Inputs: 15 | hidden_size: 2 | Loss: 0 | Scaler Type: robust | attn_dropout: 0.2 | encoder_layers: 5
Learning Rate: 0 | head_dropout: 0.2 | Dropout: 0.3 | n_head: 12 | patch_len: 3
RMSE: 0.002548342837633862 | QLIKE: 0.02295953078286472 | MAE: 0.0015394668155845233 | MAPE: 0.1552167254298638


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 42508768


Epoch: 150 | Inputs: 42 | hidden_size: 3 | Loss: 0 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 2
Learning Rate: 1 | head_dropout: 0.3 | Dropout: 0 | n_head: 15 | patch_len: 1
RMSE: 0.002526472261895962 | QLIKE: 0.02257904732598412 | MAE: 0.0015144523594421628 | MAPE: 0.15277428882298233


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 3411580


Epoch: 250 | Inputs: 5 | hidden_size: 0 | Loss: 1 | Scaler Type: robust | attn_dropout: 0.3 | encoder_layers: 1
Learning Rate: 2 | head_dropout: 0.3 | Dropout: 0.2 | n_head: 6 | patch_len: 1
RMSE: 0.002557963224595731 | QLIKE: 0.02322117251425926 | MAE: 0.0014745940763363644 | MAPE: 0.14489805479395246


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 50 | Inputs: 15 | hidden_size: 4 | Loss: 2 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 1
Learning Rate: 3 | head_dropout: 0.3 | Dropout: 0.2 | n_head: 18 | patch_len: 1
RMSE: 0.002784170765633352 | QLIKE: 0.02886488360654688 | MAE: 0.0017373163832565936 | MAPE: 0.17373532538712377


INFO:lightning_fabric.utilities.seed:Seed set to 42914685

Epoch: 150 | Inputs: 42 | hidden_size: 3 | Loss: 0 | Scaler Type: standard | attn_dropout: 0.3 | encoder_layers: 2
Learning Rate: 1 | head_dropout: 0.3 | Dropout: 0 | n_head: 15 | patch_len: 1
RMSE: 0.002526472261895962 | QLIKE: 0.02257904732598412 | MAE: 0.0015144523594421628 | MAPE: 0.15277428882298233

In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = PatchTST(h=1, input_size=42,
             hidden_size=hidden_size[3], linear_hidden_size =int(hidden_size[3]*2),
           dropout=dropout[0],  fc_dropout=dropout[0], head_dropout=0.3,
                   patch_len=patch_len[1],stride=stride[1],
           attn_dropout=0.3, n_heads=15,
             encoder_layers=2,loss=losses[0],
             max_steps=150,
           learning_rate=learning_rate[1], num_lr_decays=num_lr_decays[1],
             val_check_steps=500, batch_size=32,
             scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 2825471


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 112595934


Random Seed: 2825471
RMSE: 0.0025285294480888107 | QLIKE: 0.022681377058054647 | MAE: 0.001510341157252233 | MAPE: 0.15158276711535756


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 92226416


Random Seed: 112595934
RMSE: 0.002526389598098694 | QLIKE: 0.022537069197500037 | MAE: 0.0015046624382103755 | MAPE: 0.1517618934642381


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 50733510


Random Seed: 92226416
RMSE: 0.0025356208284881988 | QLIKE: 0.022639888991825483 | MAE: 0.0015036137105926837 | MAPE: 0.15138027661076883


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 48545993


Random Seed: 50733510
RMSE: 0.0025357141032409745 | QLIKE: 0.022630032279631604 | MAE: 0.001506830708252451 | MAPE: 0.15077532138964028


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 48545993
RMSE: 0.0025420640429757863 | QLIKE: 0.022731880607510555 | MAE: 0.0015099478295317803 | MAPE: 0.15217609757008171


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = PatchTST(h=1, input_size=42,
             hidden_size=hidden_size[3], linear_hidden_size =int(hidden_size[3]*2),
           dropout=dropout[0],  fc_dropout=dropout[0], head_dropout=0.3,
                   patch_len=patch_len[1],stride=stride[1],
           attn_dropout=0.3, n_heads=15,
             encoder_layers=2,loss=losses[0],
             max_steps=150,
           learning_rate=learning_rate[1], num_lr_decays=num_lr_decays[1],
             val_check_steps=500, batch_size=32,
             scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 119221403


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 87588445


Random Seed: 119221403
RMSE: 0.0025349559900825656 | QLIKE: 0.022669977224170558 | MAE: 0.0014912310740273 | MAPE: 0.14930031432656354


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 96541022


Random Seed: 87588445
RMSE: 0.0025269645410915067 | QLIKE: 0.02252587294335262 | MAE: 0.0015023410153186265 | MAPE: 0.15079864952497463


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 7611518


Random Seed: 96541022
RMSE: 0.0025338257028293997 | QLIKE: 0.022727550346982997 | MAE: 0.0015211734617650218 | MAPE: 0.15264523284983095


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 73594081


Random Seed: 7611518
RMSE: 0.0025313475903442546 | QLIKE: 0.022660376462104063 | MAE: 0.0014806818391207208 | MAPE: 0.14796621783062308


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 73594081
RMSE: 0.002526552682903103 | QLIKE: 0.022625196468704412 | MAE: 0.0015188277133975328 | MAPE: 0.15236499767648068


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = PatchTST(h=1, input_size=42,
             hidden_size=hidden_size[3], linear_hidden_size =int(hidden_size[3]*2),
           dropout=dropout[0],  fc_dropout=dropout[0], head_dropout=0.3,
                   patch_len=patch_len[1],stride=stride[1],
           attn_dropout=0.3, n_heads=15,
             encoder_layers=2,loss=losses[0],
             max_steps=150,
           learning_rate=learning_rate[1], num_lr_decays=num_lr_decays[1],
             val_check_steps=500, batch_size=32,
             scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 68632218


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 2424464


Random Seed: 68632218
RMSE: 0.0025359238649695953 | QLIKE: 0.02263215802339336 | MAE: 0.001521237317856245 | MAPE: 0.15253472052500797


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 79482768


Random Seed: 2424464
RMSE: 0.002542593086404771 | QLIKE: 0.02275148795152159 | MAE: 0.001503566908772015 | MAPE: 0.15081835245520525


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 17713155


Random Seed: 79482768
RMSE: 0.002530569771923122 | QLIKE: 0.022613741426567243 | MAE: 0.0015040470616405597 | MAPE: 0.15079429101861264


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 39512569


Random Seed: 17713155
RMSE: 0.002525125558291193 | QLIKE: 0.02258081227269778 | MAE: 0.0015131776400197751 | MAPE: 0.15159308930488172


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 39512569
RMSE: 0.0025272656801256415 | QLIKE: 0.02257539502081998 | MAE: 0.00151464547241268 | MAPE: 0.15199847083726345


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = PatchTST(h=1, input_size=42,
             hidden_size=hidden_size[3], linear_hidden_size =int(hidden_size[3]*2),
           dropout=dropout[0],  fc_dropout=dropout[0], head_dropout=0.3,
                   patch_len=patch_len[1],stride=stride[1],
           attn_dropout=0.3, n_heads=15,
             encoder_layers=2,loss=losses[0],
             max_steps=150,
           learning_rate=learning_rate[1], num_lr_decays=num_lr_decays[1],
             val_check_steps=500, batch_size=32,
             scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "PatchTST-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["PatchTST"].values
  else:
    Y_hat=forecasts["PatchTST-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 38034720


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 86668740


Random Seed: 38034720
RMSE: 0.0025444403199030635 | QLIKE: 0.022725880606105506 | MAE: 0.0014846430339875234 | MAPE: 0.14931496515403472


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 106623772


Random Seed: 86668740
RMSE: 0.0025228226648797516 | QLIKE: 0.022505092421147104 | MAE: 0.0014866144037495867 | MAPE: 0.14919492272765947


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 16047667


Random Seed: 106623772
RMSE: 0.0025393708802710006 | QLIKE: 0.02275585818816267 | MAE: 0.001529075465639712 | MAPE: 0.1538637773025191


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 96692183


Random Seed: 16047667
RMSE: 0.0025358349528986756 | QLIKE: 0.02268838958693529 | MAE: 0.001495735147036871 | MAPE: 0.15014735351418726


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 96692183
RMSE: 0.002534576425855726 | QLIKE: 0.022562283553832167 | MAE: 0.0015023772234653348 | MAPE: 0.15010754656635802


# Test

In [ ]:
test_length=int(len(df['NKE'])*0.3)
df1=df.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

# H=1

In [ ]:
model = PatchTST(h=1, input_size=42,
             hidden_size=hidden_size[3], linear_hidden_size =int(hidden_size[3]*2),
           dropout=dropout[0],  fc_dropout=dropout[0], head_dropout=0.3,
                   patch_len=patch_len[1],stride=stride[1],
           attn_dropout=0.3, n_heads=15,
             encoder_layers=2,loss=losses[0],
             max_steps=150,
           learning_rate=learning_rate[1], num_lr_decays=num_lr_decays[1],
             val_check_steps=500, batch_size=32,
             scaler_type='standard',
              random_seed=86668740)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=1,static_df=df_static,n_windows=None, test_size=test_length-test_length%1,step_size=1)
forecasts = forecasts.dropna()
if "PatchTST-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["PatchTST"].values
else:
  Y_hat=forecasts["PatchTST-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 86668740


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.0038693403515680717 | QLIKE: 0.03096300015291199 | MAE: 0.002454087556661601 | MAPE: 0.19106431440025265


In [ ]:
if "PatchTST-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","PatchTST-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

In [ ]:
df2.to_excel("PatchTST forecasts.xlsx")

# H=5

In [ ]:
H=5
model = PatchTST(h=H, input_size=42,
             hidden_size=hidden_size[3], linear_hidden_size =int(hidden_size[3]*2),
           dropout=dropout[0],  fc_dropout=dropout[0], head_dropout=0.3,
                   patch_len=patch_len[1],stride=stride[1],
           attn_dropout=0.3, n_heads=15,
             encoder_layers=2,loss=losses[0],
             max_steps=150,
           learning_rate=learning_rate[1], num_lr_decays=num_lr_decays[1],
             val_check_steps=500, batch_size=32,
             scaler_type='standard',
              random_seed=86668740)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=H,static_df=df_static,n_windows=None, test_size=test_length-test_length%H,step_size=H)
forecasts = forecasts.dropna()
if "PatchTST-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["PatchTST"].values
else:
  Y_hat=forecasts["PatchTST-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

if "PatchTST-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","PatchTST-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

df2.to_excel("PatchTST forecasts.xlsx")

INFO:lightning_fabric.utilities.seed:Seed set to 86668740


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.004646340034091326 | QLIKE: 0.04348540255877384 | MAE: 0.0029313563208999472 | MAPE: 0.22859590782445732


# H=10

In [ ]:
H=10
model = PatchTST(h=H, input_size=42,
             hidden_size=hidden_size[3], linear_hidden_size =int(hidden_size[3]*2),
           dropout=dropout[0],  fc_dropout=dropout[0], head_dropout=0.3,
                   patch_len=patch_len[1],stride=stride[1],
           attn_dropout=0.3, n_heads=15,
             encoder_layers=2,loss=losses[0],
             max_steps=150,
           learning_rate=learning_rate[1], num_lr_decays=num_lr_decays[1],
             val_check_steps=500, batch_size=32,
             scaler_type='standard',
              random_seed=86668740)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=H,static_df=df_static,n_windows=None, test_size=test_length-test_length%H,step_size=H)
forecasts = forecasts.dropna()
if "PatchTST-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["PatchTST"].values
else:
  Y_hat=forecasts["PatchTST-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

if "PatchTST-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","PatchTST-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

df2.to_excel("PatchTST forecasts.xlsx")

INFO:lightning_fabric.utilities.seed:Seed set to 86668740


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.005477468040900361 | QLIKE: 0.05594005733238158 | MAE: 0.0033761315431171408 | MAPE: 0.25713652533486403
